pkg/cmd/server/origin/handlers_test.go

```diff
@@ -0,0 +1,245 @@
+	newerOADMKubeResources               = "oadm/v1.2.1 (linux/amd64) kubernetes/bc4550d"
+	newerOADMOriginResources             = "oadm/v1.1.4 (linux/amd64) openshift/b348c2f"
+	newerVersionUserAgents               = []string{
+		newerOCKubeResources, newerOCOriginResources, newerOpenshiftKubectlKubeResources, newerOpenshiftKubectlOriginResources, newerOADMKubeResources, newerOADMOriginResources}
+
+	notOCVersion = "something else"
+
+	openshiftServerVersion = version.Info{GitVersion: "v1.1.3"}
+	kubeServerVersion      = kversion.Info{GitVersion: "v1.2.0"}
+)
+
+type versionSkewTestCase struct {
+	name           string
+	userAgents     []string
+	failureMessage string
+	methods        []string
+}
+
+func (tc versionSkewTestCase) Run(url string, t *testing.T) {
+	// gets always succeed
+	for _, userAgent := range tc.userAgents {
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			t.Errorf("%s: unexpected error: %v", tc.name, err)
+			return
+		}
+		req.Header.Add("User-Agent", userAgent)
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			t.Errorf("%s: unexpected error: %v", tc.name, err)
+			return
+		}
+		if resp.StatusCode != http.StatusOK {
+			t.Errorf("%s: unexpected status: %v", tc.name, resp.StatusCode)
+			return
+		}
+	}
+
+	for _, method := range tc.methods {
+		for _, userAgent := range tc.userAgents {
+			req, err := http.NewRequest(method, url, nil)
+			if err != nil {
+				t.Errorf("%s: unexpected error: %v", tc.name, err)
+				return
+			}
+			req.Header.Add("User-Agent", userAgent)
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				t.Errorf("%s: unexpected error: %v", tc.name, err)
+				return
+			}
+			if len(tc.failureMessage) == 0 {
+				if resp.StatusCode != http.StatusOK {
+					t.Errorf("%s: unexpected status: %v", tc.name, resp.StatusCode)
+					return
+				}
+
+			} else {
+				if resp.StatusCode != http.StatusForbidden {
+					t.Errorf("%s: unexpected status: %v", tc.name, resp.StatusCode)
+					return
+				}
+
+				body, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					t.Errorf("%s: unexpected error: %v", tc.name, err)
+					return
+				}
+
+				if !strings.Contains(string(body), tc.failureMessage) {
+					t.Errorf("%s: expected %v, got %v", tc.name, tc.failureMessage, string(body))
+					return
+				}
+			}
+		}
+	}
+
+}
+
+func TestVersionSkewFilterAllowAll(t *testing.T) {
+	verbs := []string{"PUT", "POST"}
+	doNothingHandler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+	})
+	config := MasterConfig{}
+	config.Options.PolicyConfig.LegacyClientPolicyConfig.LegacyClientPolicy = configapi.AllowAll
+	config.Options.PolicyConfig.LegacyClientPolicyConfig.RestrictedHTTPVerbs = verbs
+	server := httptest.NewServer(config.versionSkewFilter(openshiftServerVersion, kubeServerVersion, doNothingHandler))
+	defer server.Close()
+
+	testCases := []versionSkewTestCase{
+		{
+			name:    "missing",
+			methods: verbs,
```

Comment:
The `missing` cases in the tests need to have a slice with an empty agent to have something to `range` over in order to invoke the tests.